src/Uno.UI/UI/Xaml/Controls/Popup/Popup.Base.cs

```diff
@@ -154,26 +157,23 @@ protected internal override void OnTemplatedParentChanged(DependencyPropertyChan
 
 	private void UpdateDataContext(DependencyPropertyChangedEventArgs e)
 	{
-		_childHasOwnDataContext = false;
 		if (PropagatesDataContextToChild && Child is IDependencyObjectStoreProvider provider)
 		{
-			var dataContextProperty = provider.Store.ReadLocalValue(provider.Store.DataContextProperty);
+			var childLocalValue = provider.Store.ReadLocalValue(provider.Store.DataContextProperty);
+			_childHasOwnDataContext = childLocalValue != null && childLocalValue != DependencyProperty.UnsetValue;
+
+			var childPopupValue = provider.Store.GetValue(provider.Store.DataContextProperty, DependencyPropertyValuePrecedences.Popup);
 
-			var shouldClearValue = e != null && e.NewValue == null && dataContextProperty == e.OldValue;
+			var shouldClearValue = e != null && e.NewValue == null && childPopupValue == e.OldValue;
 			if (shouldClearValue)
 			{
 				//In this case we are clearing the DataContext that was previously set by the Popup
 				//This usually occurs when the owner of the Popup is removed from the Visual Tree
-				provider.Store.ClearValue(provider.Store.DataContextProperty, DependencyPropertyValuePrecedences.Local);
-			}
-			else if (dataContextProperty != null && dataContextProperty != DependencyProperty.UnsetValue)
-			{
-				// Child already has locally set DataContext, we shouldn't overwrite it.
-				_childHasOwnDataContext = true;
+				provider.Store.ClearValue(provider.Store.DataContextProperty, DependencyPropertyValuePrecedences.Popup);
 			}
 			else
 			{
-				provider.Store.SetValue(provider.Store.DataContextProperty, this.DataContext, DependencyPropertyValuePrecedences.Local);
+				provider.Store.SetValue(provider.Store.DataContextProperty, this.DataContext, DependencyPropertyValuePrecedences.Popup);
```

Comment:
I'm wondering if adding `FrameworkPropertyMetadataOptions.LogicalChild` to `Popup.ChildProperty` would make the inheritance precedence work seamlessly?

Reply:
Didn't work. The flyout becomes RootVisual's child later, so it inherits a null datacontext. Further attempts at playing with `ValueDoesNotInheritDataContext` also didn't work